### PR TITLE
Feature .Net v2

### PR DESCRIPTION
**Pull request by andrueastman**
This PR encompasses work for v2 of dotnet.

Related issues:-
- https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/80 (System.Text.Json migration)
- https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/201 (Add methods for GraphResponse)

### CODE REVIEW

File: Templates/templates/CSharp/Base/EntityRequest.Base.template.tt
Comment (baywet) [inline Templates/templates/CSharp/Base/EntityRequest.Base.template.tt:170]: isn't flushing the writer going to flush the underlying stream as well? https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonwriter.flush?view=net-5.0#remarks

Comment (andrueastman) [reply Templates/templates/CSharp/Base/EntityRequest.Base.template.tt:170]: Thanks for pointing this out. Will action and resolve.

**Comment by andrueastman** [comment]
Update sample generation https://github.com/microsoftgraph/msgraph-sdk-dotnet/commit/0cc99b22d741d6e4c7cbdc15338e28fcbb09a6a0

**Comment by andrueastman** [comment]
TODO 
- [ ] Merge in changes from https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/316 once its merged in

**Review by baywet — DISMISSED** [review]

**Comment by andrueastman** [comment]
@baywet Thanks for making time to have a look at this

**Review by baywet — DISMISSED** [review]

**Comment by andrueastman** [comment]
@baywet @MIchaelMainer Any chance I can get a re-review? 
The votes got reset and I would like to merge this in before https://github.com/microsoftgraph/msgraph-sdk-raptor/pull/167 so as not affect raptor tests.

**Comment by andrueastman** [comment]
Just to clarify, https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/pull/321 and https://github.com/microsoftgraph/msgraph-sdk-dotnet/pull/1058 will have to be merged in first before this is merged in.

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:359]: could casing differences derail this first comparison?

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:365]: same casing comment

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (baywet) [inline src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:382]:
```suggestion
            return entityOrComplexType?.IsBaseReferencedAsPropertyType() ?? entityOrComplexType?.Base?.IsBaseTypeReferenced() ?? false;
```

and remove the rest :)

File: Templates/templates/CSharp/Requests/MethodCollectionResponse.cs.tt
Comment (baywet) [inline Templates/templates/CSharp/Requests/MethodCollectionResponse.cs.tt:46]: nit: indent

File: Templates/templates/CSharp/Requests/IStreamRequest.cs.tt
Comment (baywet) [inline Templates/templates/CSharp/Requests/IStreamRequest.cs.tt:49]: delete this interface?

File: Templates/templates/CSharp/Requests/IMethodRequest.cs.tt
Comment (baywet) [inline Templates/templates/CSharp/Requests/IMethodRequest.cs.tt:61]: delete this interface?

File: Templates/templates/CSharp/Requests/IEntityWithReferenceRequest.cs.tt
Comment (baywet) [inline Templates/templates/CSharp/Requests/IEntityWithReferenceRequest.cs.tt:22]: delete the interface

File: Templates/templates/CSharp/Requests/EntityCollectionWithReferencesResponse.cs.tt
Comment (baywet) [inline Templates/templates/CSharp/Requests/EntityCollectionWithReferencesResponse.cs.tt:36]: nit: indent

File: Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt
Comment (baywet) [inline Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt:36]: can we have everything that calls stringBuilder.Append etc moved into a CS file please? for maintainability and ease of review.

**Review by baywet — CHANGES_REQUESTED** [review]

**Comment by MIchaelMainer** [comment]
Will need to review locally. Let's make sure we have an archive branch for the old version of the .NET templates.

**Review by MIchaelMainer — APPROVED** [review]

File: test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java
Comment (andrueastman) [reply test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java:1]: This was caused by adding a test that updated the testMetadata file by adding a new test property. This caused the java generator to create the new file.

File: test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/model/Group.cs
Comment (andrueastman) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/model/Group.cs:28]: This is addressed in the upgrade guide [here](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/feature/2.0/docs/upgrade-to-v4.md#odatatype-is-no-longer-specified-by-default-for-all-types).
Essentially, we set the odata type when
- When type derives from an abstract type
- When one of its base types is referenced as the type for a property in another type (except if the base is entity).
- When one of its base types is referenced as the type in an odata action in another type (except if the base is entity).

In this case, plannerGroup directly inherits from entity in the test metadata. So the property will not be set.

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:365]: Agreed. Will action and resolve.

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:359]: Agreed. Will action and resolve.

File: test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java
Comment (baywet) [reply test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java:1]: would you mind rebasing and kicking a new generation process for the java beta/v1 test files please? I've made a few modifications while you were working on this feature branch and I'm worried we'll get integration issues once it's merged. (the new files you added won't have the changes I've made)

File: test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/model/Group.cs
Comment (baywet) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/model/Group.cs:28]: thanks for clarifying. man those oData quirks...

File: Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt
Comment (baywet) [reply Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt:36]: it just makes it hard to review for us right now. I'd like things to be moved, but if this is too much of a burden leave it. Your call @andrueastman 😊

File: test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java
Comment (andrueastman) [reply test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java:1]: Definitely! 
I still need to test out the feedback changes on this PR but will definitely include this as well in the next update.

**Comment by baywet** [comment]
@MIchaelMainer I had nothing to do with that branch besides suggesting it to @andrueastman and making the last commit before the branch was created. Credit goes to Andrew :)

File: test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java
Comment (andrueastman) [reply test/Typewriter.Test/TestDataJavaBeta/com/microsoft/graph/requests/WorkbookChartRequestBuilder.java:1]: Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/562

File: test/Typewriter.Test/TestDataCSharp/com/microsoft/graph2/callrecords/requests/SingletonEntity1Request.cs
Comment (andrueastman) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph2/callrecords/requests/SingletonEntity1Request.cs:44]: Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/562

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:363]: Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/562

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:369]: Resolved via #562

File: Templates/templates/CSharp/Requests/MethodCollectionResponse.cs.tt
Comment (andrueastman) [reply Templates/templates/CSharp/Requests/MethodCollectionResponse.cs.tt:46]: Resolved via #562

File: src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs
Comment (andrueastman) [reply src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs:386]: Resolved via #562

File: Templates/templates/CSharp/Requests/EntityCollectionWithReferencesResponse.cs.tt
Comment (andrueastman) [reply Templates/templates/CSharp/Requests/EntityCollectionWithReferencesResponse.cs.tt:36]: Resolved via #562

File: Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt
Comment (andrueastman) [reply Templates/templates/CSharp/Base/ICollectionRequest.Base.template.tt:36]: I will leave it for now.

File: .azure-pipelines/generation-templates/dotnet.yml
Comment (andrueastman) [reply .azure-pipelines/generation-templates/dotnet.yml:23]: Yes. 
This is temporary for now as System.Text.Json version 5.0.1(which we use) is unable to be resolved by the current version of PowerShell Core which uses version 5.0.0.

File: test/Typewriter.Test/TestDataCSharp/com/microsoft/graph2/callrecords/requests/ISingletonEntity1Request.cs
Comment (andrueastman) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph2/callrecords/requests/ISingletonEntity1Request.cs:27]: We resolved to drop the rest of the interfaces when we moved to Kiota and only drop `IGraphServiceClient` for now as that changed with metadata changes. 
Dropping more interfaces would also make testing for the harder for SDK users as seen from feedback after dropping `IGraphServiceClient`. 
Even with virtual methods/properties, the setup work for testing is increased (as you would still have to create more mocks for the dependencies of the object) and hinders the developer experience.

File: test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/requests/DirectoryObjectReferenceRequest.cs
Comment (andrueastman) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/requests/DirectoryObjectReferenceRequest.cs:94]: Thanks for pointing this out. This can actually be simplified using [the existing type in core](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dev/src/Microsoft.Graph.Core/Models/ReferenceRequestBody.cs) to 

```cs
this.Method = HttpMethods.PUT;
this.ContentType = CoreConstants.MimeTypeNames.Application.Json;
var referenceRequestBody = new ReferenceRequestBody()
{
    ODataId = string.Format(@"{0}/users/{1}", this.Client.BaseUrl, id)
};
await this.SendAsync(referenceRequestBody, cancellationToken).ConfigureAwait(false);
```

Will resolve this in the next update to #562

Comment (andrueastman) [reply test/Typewriter.Test/TestDataCSharp/com/microsoft/graph/requests/DirectoryObjectReferenceRequest.cs]: Resolved via #562

**Review by baywet — APPROVED** [review]

**Pull request merged**
